File: tests/operator-summary.test.ts
/**
 * operatorSummary / withoutOperator — the sidebar's pinned Helm section.
 */
import { describe, it, expect } from 'vitest';
import { operatorSummary, withoutOperator } from '../renderer/operator-summary';
import { buildFlatNavList, buildSessionGroups, findNavIndexBySessionId } from '../renderer/session-groups';

const worker = { id: 'w1', name: 'worker', cliType: 'claude-code', workingDir: 'X:/a' } as any;
const operator = { id: 'op', name: 'Helm', cliType: 'claude-code', workingDir: 'X:/a', role: 'operator' } as any;

describe('operatorSummary', () => {
  it('is off when no session has the operator role', () => {
    expect(operatorSummary([worker], new Map(), [])).toEqual({ kind: 'off' });
  });

  it('reports the operator, its activity and the newest Helm line', () => {
    const summary = operatorSummary([worker, operator], new Map([['op', 'active']]), [
      { from: 'helm', text: 'old' }, { from: 'helm', text: 'newest' }, { from: 'you', text: 'hi' },
    ]);
    expect(summary).toEqual({ kind: 'on', id: 'op', activityLevel: 'active', lastReply: 'newest' });
  });

  it('defaults to idle with no reply yet', () => {
    expect(operatorSummary([operator], new Map(), [])).toEqual({ kind: 'on', id: 'op', activityLevel: 'idle', lastReply: null });
  });
});

describe('withoutOperator', () => {
  it('keeps the operator out of the regular session groups', () => {
    expect(withoutOperator([worker, operator])).toEqual([worker]);
    const groups = buildSessionGroups([worker, operator], () => 'X:/a', { order: [], collapsed: [] }, []);
    expect(groups.flatMap(g => g.sessions).map(s => s.id)).toEqual(['w1']);
  });
});

describe('operator as a gamepad nav target', () => {
  it('is the pinned first nav item and found by session id', () => {
    const groups = buildSessionGroups([worker, operator], () => 'X:/a', { order: [], collapsed: [] }, []);
    const nav = buildFlatNavList(groups, 'op');
    expect(nav[0]).toEqual({ type: 'operator', id: 'op', groupIndex: -1 });
    expect(nav.map(item => item.type)).toEqual(['operator', 'group-header', 'session-card']);
    expect(findNavIndexBySessionId(nav, 'op')).toBe(0);
  });

  it('is absent when the operator is off', () => {
    const groups = buildSessionGroups([worker], () => 'X:/a', { order: [], collapsed: [] }, []);
    expect(buildFlatNavList(groups, null).some(item => item.type === 'operator')).toBe(false);
  });
});

File: android/app/src/test/kotlin/com/potatomotato/helm/ui/operator/OperatorSummaryTest.kt
package com.potatomotato.helm.ui.operator

import com.potatomotato.helm.data.ChatMessage
import com.potatomotato.helm.data.HelmSession
import com.potatomotato.helm.ui.components.SessionState
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class OperatorSummaryTest {
    private fun session(id: String, role: String? = null, activity: SessionState = SessionState.Idle) = HelmSession(
        id = id,
        name = id,
        projectPath = "x:/p",
        cliType = "claudecode",
        cliTypeName = "Claude Code",
        activity = activity,
        role = role,
    )

    private fun msg(key: String, text: String, at: Long, fromPhone: Boolean = false) =
        ChatMessage(key = key, text = text, at = at, fromPhone = fromPhone)

    @Test
    fun `the operator's latest reply is the snippet, not the phone's own send`() {
        val sessions = listOf(session("a"), session("helm", role = "operator", activity = SessionState.Active))
        val threads = mapOf(
            "helm" to listOf(msg("1", "older", 1), msg("2", "newest", 2), msg("3", "my question", 3, fromPhone = true)),
            "a" to listOf(msg("9", "not the operator", 9)),
        )
        assertEquals(
            OperatorSummary.On(id = "helm", activity = SessionState.Active, lastReply = "newest"),
            operatorSummary(sessions, threads),
        )
    }

    @Test
    fun `an operator with no replies yet has no snippet`() {
        val summary = operatorSummary(listOf(session("helm", role = "operator")), emptyMap())
        assertNull((summary as OperatorSummary.On).lastReply)
    }

    @Test
    fun `no operator session means the section is off`() {
        assertEquals(OperatorSummary.Off, operatorSummary(listOf(session("a")), emptyMap()))
    }

    @Test
    fun `the operator is excluded from the regular session list`() {
        val sessions = listOf(session("a"), session("helm", role = "operator"), session("b"))
        assertEquals(listOf("a", "b"), withoutOperator(sessions).map { it.id })
    }
}
